cirq/sim/wave_function.py

```diff
@@ -15,17 +15,131 @@
 
 import itertools
 
-from typing import Iterable, List, Sequence, Tuple, Type, Union, TYPE_CHECKING
+from typing import (Iterable, List, Sequence, Tuple, Type, Union, TYPE_CHECKING,
+                    Optional, Dict)
 
+import abc
 import numpy as np
 
-from cirq import linalg
+from cirq import linalg, ops
 
 if TYPE_CHECKING:
     # pylint: disable=unused-import
     from cirq.sim import simulator
 
 
+class State:
+    """Represent a quantum state (wave function), and provide various
+    convenience methods.
+
+    Attributes:
+        qubit_map: A map from the Qubits in the Circuit to the the index
+            of this qubit for a canonical ordering. This canonical ordering is
+            used to define the state (see the state() method).
+    """
+
+    def __init__(self, qubit_map: Optional[Dict[ops.QubitId, int]]):
+        self.qubit_map = qubit_map or {}
+
+    @abc.abstractmethod
+    def state(self) -> np.ndarray:
+        """Return the state (wave function).
+
+        The state is returned in the computational basis with these basis
+        states defined by the `qubit_map`. In particular the value in the
+        `qubit_map` is the index of the qubit, and these are translated into
+        binary vectors where the last qubit is the 1s bit of the index, the
+        second-to-last is the 2s bit of the index, and so forth (i.e. big
+        endian ordering).
+
+        Example:
+             qubit_map: {QubitA: 0, QubitB: 1, QubitC: 2}
+             Then the returned vector will have indices mapped to qubit basis
+             states like the following table
+
+                    | QubitA | QubitB | QubitC
+                :-: | :----: | :----: | :----:
+                 0  |   0    |   0    |   0
+                 1  |   0    |   0    |   1
+                 2  |   0    |   1    |   0
+                 3  |   0    |   1    |   1
+                 4  |   1    |   0    |   0
+                 5  |   1    |   0    |   1
+                 6  |   1    |   1    |   0
+                 7  |   1    |   1    |   1
+
+        """
+        raise NotImplementedError()
+
+    def dirac_notation(self, decimals: int = 2) -> str:
+        """Returns the state as a string in Dirac notation.
+
+        Args:
+            decimals: How many decimals to include in the pretty print.
+
+        Returns:
+            A pretty string consisting of a sum of computational basis kets
+            and non-zero floats of the specified accuracy."""
+        return dirac_notation(self.state(), decimals)
+
+    def density_matrix(self, qubits: List[ops.QubitId] = None) -> np.ndarray:
```

Comment:
I would rename this to `density_matrix_of`, because it's specific to the given qubits instead of the entire state.